File: 013CaesarCipherEncryptor/caesarCipherEncryptor.py
# Solution 1
# O(n) time / O(n) space

def caesarCipherEncryptor(string, key):
    newLetters = []
    newKey = key % 26
    for letter in string:
        newLetters.append(getNewLetter(letter, newKey))
    return ''.join(newLetters)


def getNewLetter(letter, key):
    getNewLetterCode = ord(letter) + key
    return chr(getNewLetterCode) if getNewLetterCode <= 122 else chr(96 + getNewLetterCode % 122)

# Solution 2
# O(n) time / O(n) space


def caesarCipherEncryptor(string, key):
    newLetters = []
    newKey = key % 26
    alphabet = list('abcdefghijklmnopqrstuvwxyz')
    for letter in string:
        newLetters.append(getNewLetter(letter, newKey, alphabet))
    return ''.join(newLetters)


def getNewLetter(letter, key, alphabet):
    newLetterCode = alphabet.index(letter) + key
    return alphabet[newLetterCode % 26]

File: 013CaesarCipherEncryptor/caesarCipherEncryptor.js
// Solution 1
// O(n) time / O(n) space

const caesarCipherEncryptor = (string, key) => {
  const newLetters = [];
  const newKey = key % 26;
  for (const letter of string) {
    newLetters.push(getNewLetter(letter, newKey));
  }
  return newLetters.join('');
};

const getNewLetter = (letter, key) => {
  const newLetterCode = letter.charCodeAt() + key;
  return newLetterCode <= 122
    ? String.fromCharCode(newLetterCode)
    : String.fromCharCode(96 + (newLetterCode % 122));
};

// Solution 2
// O(n) time / O(n) space

const caesarCipherEncryptor = (string, key) => {
  const newLetters = [];
  const newKey = key % 26;
  const alphabet = 'abcdefghijklmnopqrstuvwxyz';
  for (const letter of string) {
    newLetters.push(getNewLetter(letter, newKey, alphabet));
  }
  return newLetters.join('');
};

const getNewLetter = (letter, key, alphabet) => {
  const newLetterCode = alphabet.indexOf(letter) + key;
  return alphabet[newLetterCode] % 26;
};
